src/synapse_api.py

```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Helper module used to manage interactions with Synapse API."""
+
+# pylint: disable=too-few-public-methods, too-many-arguments
+
+import hashlib
+import hmac
+import logging
+import typing
+
+import ops
+import requests
+from ops.charm import CharmBase
+
+from constants import SYNAPSE_CONTAINER_NAME
+
+logger = logging.getLogger(__name__)
+
+SYNAPSE_URL = "http://localhost:8008"
+URL_REGISTER = f"{SYNAPSE_URL}/_synapse/admin/v1/register"
+
+
+class RegisterUserError(Exception):
+    """Exception raised when registering user via API fails.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the RegisterUserError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class NetworkError(Exception):
+    """Exception raised when requesting API fails due network issues.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the NetworkError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class SynapseAPI:
+    """The Synapse API handler."""
+
+    def __init__(self, charm: CharmBase):
```

Comment:
What is the reason for needing a class here please?

Reply:
This way register_user and future methods wouldn't need charm or synapse as parameters. Also makes it easier to define some methods related only to the API private. I believe there is a similar behavior when another charms as Prometheus and Grafana.